[SPARK-32187][PYTHON][DOCS] Doc on Python packaging (#29806)
apache/spark: python/docs/source/user_guide/python_packaging.rst

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

=========================
3rd Party Python Packages
=========================

When you want to run your PySpark application on a cluster such as YARN, Kubernetes, Mesos, etc., you need to make

> sure that the your code and all used libraries are available on the executors.

Reviewer: s/the//

> sure that the your code and all used libraries are available on the executors.

Reviewer 2: ```suggestion
sure that your code and all used libraries are available on the executors.
```

As an example let's say you may want to run the `Pandas UDF's examples <arrow_pandas.rst#series-to-scalar>`_.

> As it uses pyarrow as an underlying implementation we need to make sure to have pcyarrow installed on each executor

Reviewer 2: ```suggestion
As it uses pyarrow as an underlying implementation we need to make sure to have pyarrow installed on each executor
```

            [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
            ("id", "v"))

        @pandas_udf("double")
        def mean_udf(v: pd.Series) -> float:
            return v.mean()

        print(df.groupby("id").agg(mean_udf(df['v'])).collect())

    if __name__ == "__main__":
        main(SparkSession.builder.getOrCreate())

There are multiple ways to ship the dependencies to the cluster:

- Using PySpark Native Features
- Using Zipped Virtual Environment
- Using PEX

Using PySpark Native Features
-----------------------------

PySpark allows to upload Python files (``.py``), zipped Python packages (``.zip``), and Egg files (``.egg``)
to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly
calling :meth:`pyspark.SparkContext.addPyFile`.

This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole
packages and upload them. Using :meth:`pyspark.SparkContext.addPyFile` allows to upload code
even after having started your job.

Note that it doesn't allow to add packages built as `Wheels <https://www.python.org/dev/peps/pep-0427/>`_ and therefore doesn't

> allowing to include dependencies with native code.

Reviewer: s/allowing/allow/?

> allowing to include dependencies with native code.

Reviewer 2: ```suggestion
allow to include dependencies with native code.
```

Using Zipped Virtual Environment
--------------------------------

The idea of zipped environments is to zip your whole `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_, 

> ship it to the cluster, unzip it remotly and target the Python interpreter from inside this zipped environment. Note that this

Reviewer: s/remotly/remotely/?

> ship it to the cluster, unzip it remotly and target the Python interpreter from inside this zipped environment. Note that this

Reviewer 2: ```suggestion
ship it to the cluster, unzip it remotely and target the Python interpreter from inside this zipped environment. Note that this
```

is currently supported *only for YARN*.

> Zip Virtual Environment

Reviewer 3: Could we emphasize here, that these solutions work only on YARN?

Reviewer 2: Actually I did above `*only for YARN*` .. :-). If that sounds not enough, I am fine by emphasizing more.

Reviewer 3: Sorry, I am thinking more about putting this note in the header or something equivalent :)

Author: Putting besides Kubernetes which is all docker based I don't see a restriction on why this couldn't work with other deployements modes. 
I know Mesos pretty well as we also have to a mesos cluster (even though I have never tried to deploy Sprak on it) and I think it will work the same. For Standalone mode I can confirm it works.

So spark-submit I'm pretty sure it works. The restriction to me is only for running SparkSession.Builder from inside Python where there is one argument missing to inject archives as config. But it is a minor fix imo. I could even do it. I would prefer not saying restricted to YARN and do the fix instead and inject archives via a config param.

Reviewer 3: In such case could we add (and validate) non-YARN example? That would be useful, especially when both `conda-pack` and   `venv-pack` already provide YARN ones.

Author: Sorry for the confusion. I had a look again and the feature that zip files are extracted on the executor really seems specific to YARN (https://github.com/apache/spark/blob/3edab6cc1d70c102093e973a2cf97208db19be8c/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala#L563. I also tried and  '--archives' actually does nothing in Standalone mode.
So the current state is good. Zippable env only supported for YARN. PEX should work everywhere as it is shipped as a file.
So I guess

~~~~~~~~~~~~~~~~~~~~~~~

You can zip the virtual environment on your own or use tools for doing this:

* `conda-pack <https://conda.github.io/conda-pack/spark.html>`_ for conda environments
* `venv-pack <https://jcristharif.com/venv-pack/spark.html>`_ for virtual environments

Example with `conda-pack`:

.. code-block:: bash

    conda create -y -n conda_env -c conda-forge \
      pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 conda-pack==0.4.0
    conda activate conda_env
    conda pack -f -o conda_env.tar.gz

Upload to Spark Executors
~~~~~~~~~~~~~~~~~~~~~~~~~

> The unzip will be done by Spark when using target ``--archives`` option in spark-submit

Reviewer: s/unzip/unzipping/?

> The unzip will be done by Spark when using target ``--archives`` option in spark-submit

Reviewer 2: ```suggestion
Unzipping will be done by Spark when using target ``--archives`` option in spark-submit
```

or setting ``spark.yarn.dist.archives`` configuration.

Example with ``spark-submit``:

.. code-block:: bash

    export PYSPARK_DRIVER_PYTHON=python
    export PYSPARK_PYTHON=./environment/bin/python
    spark-submit --master=yarn --deploy-mode client \
      --archives conda_env.tar.gz#environment app.py

Example using ``SparkSession.builder``:

.. code-block:: python

    import os
    from pyspark.sql import SparkSession
    from app import main

    os.environ['PYSPARK_PYTHON'] = "./environment/bin/python"
    builder = SparkSession.builder.master("yarn").config(
        "spark.yarn.dist.archives", "conda_env.tar.gz#environment")
    spark = builder.getOrCreate()
    main(spark)

Using PEX
---------

`PEX <https://github.com/pantsbuild/pex>`_ is a library for generating ``.pex`` (Python EXecutable) files.

> A PEX file is a self contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).

Reviewer: s/self contained/self-contained/?

> A PEX file is a self contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).

Reviewer 2: ```suggestion
A PEX file is a self-contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).
```

      ("id", "v"))

    @pandas_udf("double", PandasUDFType.GROUPED_AGG)
    def mean_udf(v: pd.Series):
      return v.mean()

    print(df.groupby("id").agg(mean_udf(df['v'])).collect())

  if __name__ == "__main__":
    spark = SparkSession.builder.getOrCreate()
    main(spark)

There are multiple ways to ship the dependencies to the cluster:

- Using py-files
- Using a zipped virtual environment
- Using PEX
- Using Docker

**************
Using py-files
**************

PySpark allows to upload python files to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly calling `addPyFile
<../reference/api/pyspark.SparkContext.addPyFile.rst>`_ on the SparkContext.

This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole packages and upload them. 
Using `addPyFile <../reference/api/pyspark.SparkContext.addPyFile.rst>`_ allows to upload code even after having started your job.

It doesn't allow to add packages built as `Wheels <https://www.python.org/dev/peps/pep-0427/>`_ and therefore doesn't allowing to include dependencies with native code.

**********************************

> Using a zipped virtual environment

Reviewer 3: If I recall correctly this can be done only with YARN, right?

Reviewer 2: Yeah, it does.

           .master("yarn") \
           .config("spark.yarn.dist.archives",
                   "conda_env.tar.gz#environment")
  spark = builder.getOrCreate()
  main(spark)

*********
Using PEX
*********

`PEX <https://github.com/pantsbuild/pex>`_ is a library for generating .pex (Python EXecutable) files.
A PEX file is a self contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (aka fat JARs).

You need to build the PEX file somewhere with all your requirements and then upload it to each Spark executor.

Using the CLI to build the PEX file
-----------------------------------

.. code-block:: bash

  pex pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 -o myarchive.pex

Invoking the pex file will by default invoke the Python interpreter. pyarrow, pandas and pyspark will be included in the pex file.

.. code-block:: bash
  
  ./myarchive.pex
  Python 3.6.6 (default, Jan 26 2019, 16:53:05)
  (InteractiveConsole)
  >>> import pyarrow
  >>> import pandas
  >>> import pyspark
  >>>

This can also be done directly with the Python API. `More infos <https://pex.readthedocs.io/en/stable/buildingpex.html>`_ on how to build PEX files.

> Upload it to the Spark executors
> --------------------------------
> 
> The upload can be done by setting ``--files`` option in spark-submit or setting ``spark.files`` configuration (``spark.yarn.dist.files`` on YARN) 
> and changing the ``PYSPARK_PYTHON`` environment variable to change the Python interpreter to the PEX executable on each executor.
> 
> Example with spark-submit on YARN:
> 
> .. code-block:: bash
> 
>   export PYSPARK_DRIVER_PYTHON=python
>   export PYSPARK_PYTHON=./myarchive.pex
>   spark-submit --master=yarn --deploy-mode client \
>   --files myarchive.pex \
>   main.py
> 
> Example using SparkSession.builder on YARN:
> 
> .. code-block:: python
> 
>   import os
>   from pyspark.sql import SparkSession
>   from main import main
>   os.environ['PYSPARK_PYTHON']="./myarchive.pex"
>   builder = SparkSession.builder
>   builder.master("yarn") \
>          .config("spark.submit.deployMode", "client") \
>          .config("spark.yarn.dist.files", "myarchive.pex")
>   spark = builder.getOrCreate()
>   main(spark)
> 
> Notes
> -----
> 
> The Python interpreter that has been used to generate the PEX file must be available on each executor. PEX doesn't include the Python interpreter.
> 
> In YARN cluster mode you may also need to set ``PYSPARK_PYTHON`` environment variable on the AppMaster ``--conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=./myarchive.pex``.
> 
> An end-to-end Docker example for deploying a standalone PySpark with ``SparkSession.builder`` and PEX
> can be found `here <https://github.com/criteo/cluster-pack/blob/master/examples/spark-with-S3/README.md>`_ 
> (it uses cluster-pack, a library on top of PEX that automatizes the the intermediate step of having to create & upload the PEX manually).
> 
> 
> **********************************
> Using Docker
> **********************************
> 
> Deploy you Spark application as usual with docker containers (on Standalone, Mesos, Kubernetes) and install all required packages globally in the Docker container.
> 
> 
> *****************************************************
> What about the Spark JARs/Python code in itsself ?
> *****************************************************
> 
> PySpark should be included in the PEX/zipped env in order to be shipped along the other libraries.
> 
> The Spark jars will be picked up:
> - from ``SPARK_HOME`` if this environment variable is set
> - from the pyspark package if ``SPARK_HOME`` is not set (it allows to consistently use same pyspark package)
> - you can also set the ``spark.jars``, ``spark.yarn.dist.jars``, ``spark.yarn.dist.archives`` parameters to make them available directly on distributed storage

Reviewer 2: ```suggestion
..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

=========================
3rd Party Python Packages
=========================

When you want to run your PySpark application on a cluster such as YARN, Kubernetes, Mesos, etc., you need to make
sure that the your code and all used libraries are available on the executors.

As an example let's say you may want to run the `Pandas UDF's examples <arrow_pandas.rst#series-to-scalar>`_.
As it uses pyarrow as an underlying implementation we need to make sure to have pcyarrow installed on each executor
on the cluster. Otherwise you may get errors such as ``ModuleNotFoundError: No module named 'pyarrow'``.

Here is the script ``app.py`` from the previous example that will be executed on the cluster:

.. code-block:: python

    import pandas as pd
    from pyspark.sql.functions import pandas_udf
    from pyspark.sql import SparkSession

    def main(spark):
        df = spark.createDataFrame(
            [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
            ("id", "v"))

        @pandas_udf("double")
        def mean_udf(v: pd.Series) -> float:
            return v.mean()

        print(df.groupby("id").agg(mean_udf(df['v'])).collect())

    if __name__ == "__main__":
        main(SparkSession.builder.getOrCreate())

There are multiple ways to ship the dependencies to the cluster:

- Using PySpark Native Features
- Using Zipped Virtual Environment
- Using PEX

Using PySpark Native Features
-----------------------------

PySpark allows to upload Python files (``.py``), zipped Python packages (``.zip``), and Egg files (``.egg``)
to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly
calling :meth:`pyspark.SparkContext.addPyFile`.

This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole
packages and upload them. Using :meth:`pyspark.SparkContext.addPyFile` allows to upload code
even after having started your job.

Note that it doesn't allow to add packages built as `Wheels <https://www.python.org/dev/peps/pep-0427/>`_ and therefore doesn't
allowing to include dependencies with native code.

Using Zipped Virtual Environment
--------------------------------

The idea of zipped environments is to zip your whole `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_, 
ship it to the cluster, unzip it remotly and target the Python interpreter from inside this zipped environment. Note that this
is currently supported *only for YARN*.

Zip Virtual Environment
~~~~~~~~~~~~~~~~~~~~~~~

You can zip the virtual environment on your own or use tools for doing this:

* `conda-pack <https://conda.github.io/conda-pack/spark.html>`_ for conda environments
* `venv-pack <https://jcristharif.com/venv-pack/spark.html>`_ for virtual environments

Example with `conda-pack`:

.. code-block:: bash

    conda create -y -n conda_env -c conda-forge \
      pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 conda-pack==0.4.0
    conda activate conda_env
    conda pack -f -o conda_env.tar.gz

Upload to Spark Executors
~~~~~~~~~~~~~~~~~~~~~~~~~

The unzip will be done by Spark when using target ``--archives`` option in spark-submit
or setting ``spark.yarn.dist.archives`` configuration.

Example with ``spark-submit``:

.. code-block:: bash

    export PYSPARK_DRIVER_PYTHON=python
    export PYSPARK_PYTHON=./environment/bin/python
    spark-submit --master=yarn --deploy-mode client \
      --archives conda_env.tar.gz#environment app.py

Example using ``SparkSession.builder``:

.. code-block:: python

    import os
    from pyspark.sql import SparkSession
    from app import main

    os.environ['PYSPARK_PYTHON'] = "./environment/bin/python"
    builder = SparkSession.builder.master("yarn").config(
        "spark.yarn.dist.archives", "conda_env.tar.gz#environment")
    spark = builder.getOrCreate()
    main(spark)

Using PEX
---------

`PEX <https://github.com/pantsbuild/pex>`_ is a library for generating ``.pex`` (Python EXecutable) files.
A PEX file is a self contained executable Python environment. It can be seen as the Python equivalent of Java uber-JARs (a.k.a. fat JARs).

You need to build the PEX file somewhere with all your requirements and then upload it to each Spark executor.

Using CLI to Build PEX file
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: bash

    pex pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 -o myarchive.pex

Invoking the PEX file will by default invoke the Python interpreter. pyarrow, pandas and pyspark will be included in the PEX file.

.. code-block:: bash

    ./myarchive.pex
    Python 3.6.6 (default, Jan 26 2019, 16:53:05)
    (InteractiveConsole)
    >>> import pyarrow
    >>> import pandas
    >>> import pyspark
    >>>

This can also be done directly with the Python API. For more information on how to build PEX files,
please refer to `Building .pex files <https://pex.readthedocs.io/en/stable/buildingpex.html>`_

Upload to Spark Executors
~~~~~~~~~~~~~~~~~~~~~~~~~

The upload can be done by setting ``--files`` option in spark-submit or setting ``spark.files`` configuration (``spark.yarn.dist.files`` on YARN) 
and changing the ``PYSPARK_PYTHON`` environment variable to change the Python interpreter to the PEX executable on each executor.

..
   TODO: we should also document the way on other cluster modes.

Example with ``spark-submit`` on YARN:

.. code-block:: bash

    export PYSPARK_DRIVER_PYTHON=python
    export PYSPARK_PYTHON=./myarchive.pex
    spark-submit --master=yarn --deploy-mode client --files myarchive.pex app.py

Example using ``SparkSession.builder`` on YARN:

.. code-block:: python

    import os
    from pyspark.sql import SparkSession
    from app import main

    os.environ['PYSPARK_PYTHON']="./myarchive.pex"
    builder = SparkSession.builder
    builder.master("yarn") \
         .config("spark.submit.deployMode", "client") \
         .config("spark.yarn.dist.files", "myarchive.pex")
    spark = builder.getOrCreate()
    main(spark)

Notes
~~~~~

* The Python interpreter that has been used to generate the PEX file must be available on each executor. PEX doesn't include the Python interpreter.

* In YARN cluster mode you may also need to set ``PYSPARK_PYTHON`` environment variable on the AppMaster ``--conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=./myarchive.pex``.

* An end-to-end Docker example for deploying a standalone PySpark with ``SparkSession.builder`` and PEX can be found `here <https://github.com/criteo/cluster-pack/blob/master/examples/spark-with-S3/README.md>`_ - it uses cluster-pack, a library on top of PEX that automatizes the the intermediate step of having to create & upload the PEX manually.
```
